src/librustc_metadata/decoder.rs

```diff
@@ -496,30 +467,106 @@ impl<'a, 'tcx> CrateMetadata {
         }
     }
 
-    pub fn item_name(&self, item_index: DefIndex) -> Symbol {
-        self.def_key(item_index)
-            .disambiguated_data
-            .data
-            .get_opt_name()
-            .expect("no name in item_name")
-            .as_symbol()
+    pub fn item_name(&self, item_index: DefIndex, sess: &Session) -> Symbol {
+        if !self.is_proc_macro(item_index, sess) {
+            self.def_key(item_index, sess)
+                .disambiguated_data
+                .data
+                .get_opt_name()
+                .expect("no name in item_name")
+                .as_symbol()
+        } else {
+            self.get_proc_macros(sess).unwrap()[&item_index].name
+        }
     }
 
-    pub fn def_kind(&self, index: DefIndex) -> Option<DefKind> {
-        if !self.is_proc_macro(index) {
+    pub fn def_kind(&self, index: DefIndex, sess: &Session) -> Option<DefKind> {
+        if !self.is_proc_macro(index, sess) {
             self.entry(index).kind.def_kind()
         } else {
             Some(DefKind::Macro(
-                self.proc_macros.as_ref().unwrap()[index.to_proc_macro_index()].1.macro_kind()
+                self.get_proc_macros(sess).unwrap()[&index].ext.macro_kind()
             ))
         }
     }
 
     pub fn get_span(&self, index: DefIndex, sess: &Session) -> Span {
-        match self.is_proc_macro(index) {
-            true => DUMMY_SP,
-            false => self.entry(index).span.decode((self, sess)),
+        self.entry(index).span.decode((self, sess))
+    }
+
+
+    pub fn get_proc_macros(&self, sess: &Session) -> Option<FxHashMap<DefIndex, FullProcMacro>>{
+        let proc_macros = self.root.proc_macro_data.as_ref().map(|proc_macros| {
+            let proc_macros = proc_macros.decode((self, sess));
+            if sess.opts.debugging_opts.dual_proc_macros {
+                let host_lib = self.host_lib.as_ref().unwrap();
+                self.load_proc_macros(
+                    &host_lib.metadata.get_root(),
+                    self.span,
+                    proc_macros,
+                    sess
+                )
+            } else {
+                self.load_proc_macros(&self.root,self.span, proc_macros, sess)
+            }
+        });
+        proc_macros
+    }
+
+    /// Loads custom derive macros.
+    ///
+    /// Note that this is intentionally similar to how we load plugins today,
+    /// but also intentionally separate. Plugins are likely always going to be
+    /// implemented as dynamic libraries, but we have a possible future where
+    /// custom derive (and other macro-1.1 style features) are implemented via
+    /// executables and custom IPC.
+    fn load_proc_macros(&self, root: &CrateRoot<'_>,
+                          span: Span,
+                          macros: impl ExactSizeIterator<Item = ProcMacroData>,
+                          sess: &Session)
+                          -> FxHashMap<DefIndex, FullProcMacro> {
```

Comment:
This function should only be used to load *one* proc macro, *not* all of them at once.

Reply:
This is a helper function called by `get_proc_macros`, which needs to load all proc macros.

If you're talking about modifying `get_proc_macros`, I'm not sure if there's much benefit to that. We need to be able to get a list of all proc macros in order to enumerate the crate DefIds. Are you worried about the performance impact?

Reply:
What I'm saying is that you shouldn't load the actual "syntax extension" object until the expansion system is asking for it.

Whereas you also use `get_proc_macros` to check whether a `DefId` is a proc macro, which should be *very cheap* and *perform no allocation*.